### ISSUE TITLE
ordering and priority of inflection tables

### ISSUE THREAD

**Issue by balmas** [alpheios-project/inflection-tables#23]
Info on requirements for ordering and priorities for showing multiple inflection tables for a form. 

In Alpheios V1 the basic logic used was this:

1. Inflection tables were offered in the same order in which the inflections appeared in the popup. This was determined by the @order attribute on the inflection inflection data returned by the morphological analysis. (In the V2 code this is currently implemented via Lexeme.getSortByTwoLemmaFeatures)

2. within an individual language, defined which tables should link to other tables:
Latin:
```
{     noun: { keys: ['noun'], links: [] },
      // link verb participles to both verb and adjective inflection tables
      adjective: { keys: ['adjective'], links: [] },
      verb_participle: { keys: ['verb_participle','adjective'], links: ['verb'] },
      supine: { keys: ['verb_supine'], links: ['verb'] },
      infinitive: { keys: ['verb_infinitive'], links: [] },
      imperative: { keys: ['verb_imperative'], links: [] },
      gerundive: { keys: ['verb_gerundive'],links: [] },
      verb_irregular: { keys: ['verb_irregular'],links: []},
      verb: { keys: ['verb'], links: [] }
};

```
Greek:
```
{     noun: { keys: ['noun'], links: [] },
      adjective: {keys: ['adjective'], links:[] },
      adjective_simplified: {keys: ['adjective_simplified'], links:[] },
      noun_simplified: {keys: ['noun_simplified'], links:[] },
      pronoun: { keys: ['pronoun'],links:[] },
      irregular: { keys: ['pronoun_interrogative'],links:[] },
      article: { keys: ['article'],links:[] },
      numeral: { keys: ['numeral'],links:[] },
      verb: { keys: ['verb'],links:['verb_all'] },
      verb_participle: { keys: ['verb_participle'],links:['verb_all'] },
      verb_all: { keys: ['verb_all'],links:['verb'] },
      adverb: {keys: ['adverb'],links:[]}
};
```
3. if an inflection table was a more granular table than just part of speech (e.g. a specific verb form), then we automatically added a link to the default table for that part of speech

**Comment by kirlat** [comment]
I'm testing a solution now and have issues with some complex situations that arise during testing. It seems to be not so easy sometimes (but maybe I over complicate this?)

One of the test cases is a Latin word `amatum`.
Morphological analyzer returns the following data on it:
```
amatum
    Lexeme 1: amo
        Inflections:
            1 - verb participle
            2 - verb participle
            3 - verb participle
            4 - verb participle
            5 - supine
    Lexeme 2: amatus
        Inflections:
            1 - adjective
            2 - adjective
            3 - adjective
            4 - adjective
    Lexeme 3: amata
        Inflections:
            1 - noun
 ```
We go lexeme after lexeme, inflection after inflection, and assemble a list of tables we need to present.

For Lexeme 1 we combine endings for all 4 verb participle inflections together and build one verb participle table from them. However, a verb participle table has a linked verb table, so we should add it too. Then we build a supine table with endings matching inflection 5.

So for Lexeme 1 we produced:
   1. Verb participle table (based on inflections 1 to 4)
   2. Verb table (linked from verb participle)
   3. Supine table (based on inflection 5)

The question is with what data should we build a verb table? It seems to be that we need to take inflections 1 to 4, replace a part of speech there from `verb participle` to `verb` and run matching against all known verb endings. Is that correct?

For Lexeme 2 we combine endings for 4 adjective inflections together and build an adjectives table from that. However, it seems that for adjectives we need to build a verb participle table too (as adjective is listed in keys of a verb participle table). And we also need to show a verb table which is linked from the verb participle one.

So for Lexeme 2 we should have three tables:
   1. Adjective table
   2. Verb participle table (should we base it on adjective inflections but replace adjective POFS with verb participle and run an ending matching process)
   3. Verb table (should we base it on adjective inflections but replace adjective POFS with verb and run an ending matching process)

Lexeme 3 is simple. We have just:
   1. Noun table

So overall we have:
   1. Verb participle table from Lexeme 1 data
   2. Verb table from Lexeme 1 data
   3. Supine table from Lexeme 1 data
   4. Adjective table from Lexeme 2 data
   5. Verb participle table from Lexeme 2 data
   6. Verb table from Lexeme 2 data
   7. Noun table from Lexeme 3 data.

The question is should we merge 1 and 5, 2 and 6 before showing? They are build for different lexemes. If we merge them, a difference between lemma data (highlightings or difference in endings if we would have any complex matching rules) will be lost. Should we do that?

We group tables by part of speech before showing. So the other question is what should be a proper order? If we follow the order in a popup where words are shown as:
```
Lemma 1: Verb
       Verb participle
       Supine
Lemma 2: Adjective
Lemma 3: Noun
```
so should we just follow an order of POFSs there (verb, verb participle, supine, adjective, noun) and group tables as:
```
Verb
   Verb table from Lexeme 1 data
   Verb table from Lexeme 2 data
Verb participle
   Verb participle table from Lexeme 1 data
   Verb participle table from Lexeme 2 data
Supine
   Supine table from Lexeme 1 data
Adjective
   Adjective table from Lexeme 2 data
Noun
  Noun table from Lexeme 3 data
```
Would that be correct? Seems a little messed up to me as it mixes data for lexemes one and two together. Or maybe we should group it by the lexeme? It would make sense, but is a deviation from a current UI:
```
Lexeme 1 (amo, verb):
   1. Verb participle table
   2. Verb table
   3. Supine table
Lexeme 2 (amatus, adjective):
   1. Adjective table
   2. Verb participle table
   3. Verb table
Lexeme 3 (amata):
   1. Noun table
```

So what would be the cleanest and most logical way to present information from the point of view of a language student? This seems to be complex, with many different ways to show data, and it would be great to establish and clear and deterministic algorithm for displaying tables, if possible, and add it to documentation, too, to avoid any dis-ambiguities in the future. What do you think? Really need your help here :-)

**Comment by balmas** [comment]
These are excellent questions.  In V1 we aggregated by table, showing all endings for all lexemes which had a possibility in that table
![screenshot from 2018-07-13 07-54-21](https://user-images.githubusercontent.com/644870/42690389-fef7564e-8671-11e8-9577-dacd7f1a8ce2.png)
![screenshot from 2018-07-13 07-54-15](https://user-images.githubusercontent.com/644870/42690390-ff0aa3fc-8671-11e8-86d4-e72f84823abe.png)
![screenshot from 2018-07-13 07-54-08](https://user-images.githubusercontent.com/644870/42690391-ff1bcbf0-8671-11e8-97c2-78384e5b4f84.png)
![screenshot from 2018-07-13 07-53-55](https://user-images.githubusercontent.com/644870/42690392-ff339fd2-8671-11e8-9f60-9068bdf079d3.png)
![screenshot from 2018-07-13 07-53-44](https://user-images.githubusercontent.com/644870/42690393-ff4fbe7e-8671-11e8-8d26-5d6b0057f383.png)

**Comment by balmas** [comment]
In answer to: "The question is with what data should we build a verb table? It seems to be that we need to take inflections 1 to 4, replace a part of speech there from verb participle to verb and run matching against all known verb endings. Is that correct?" YES, this is correct.

**Comment by balmas** [comment]
@kirlat for now I would like to take a different approach than we did in V1 and rather than aggregate the lexemes, so a distinct view of each table per lexeme, e.g. as above:

```
Lexeme 1 (amo, verb):
   1. Verb participle table
   2. Verb table
   3. Supine table
Lexeme 2 (amatus, adjective):
   1. Adjective table
   2. Verb participle table
   3. Verb table
Lexeme 3 (amata):
   1. Noun table
```

**Comment by balmas** [comment]
Comments from @monzug in #76:

 for the supine, we do not have the full irregular table generated but a complicated navigation through drop-downs. first I can chose from noun, verb participle and supine. if I chose supine, I have verb-irregular-supine and supine. I would prefer a simplify version of this navigation.
noun
verb irregular ---> verb irregular supine
verb irregular (where I can see the verb participle too)
what do you think?

And response from @kirlat 

I do not have enough knowledge to decide how it's best to present inflection data, so I'll rely on you and Bridget here. But on my opinion drop-down navigation is complicated as it's hard to see all the choices at once, especially with dependent drop-downs for POSs (i.e. verbs have its own drop-downs for verb choices, but it changes for a different POS).

So what I want to offer for discussion if maybe we can stop using drop downs all together and maybe instead show titles of collapsed inflection tables we have on the word. "Main" tables can be shown expanded. That will allow users to see at once all tables that are available, and then expands the ones of interest to compare them side by side.

That would also be better UX-wise: expanding or collapsing a table requires a single click and using drop-downs require two (one to open a drop-down and another to make a click).

Please let me know what do you think.

And further response from @monzug 
it's a good suggestion but I am not sure it would work in term of the flexibility that users can have right now. e.g. the main table has different views so users can choose the one that fits better.

**Comment by balmas** [comment]
My opinion on navigation is this:

Users must be able to do the following in response to a lookup of a word where that word has one or more inflections:

1) navigate to the appropriate main table for each distinct lexical entry in the popup
2) the order of the tables shown in that navigation should match the order of the lexical entries in the popup
3) if a main table for a lexical entry is one which has other associated tables (such as a simplified table, or an expanded table, or a regular table in the case of an irregular verb, etc. ) then ideally that would be offered as secondary navigation only once the user has selected the table that is linked to the lexical entry in the popup, and not before

And once https://github.com/alpheios-project/components/issues/40 is fixed, we should always be able to navigate to a display which allows the user to pick and choose from the full set of available inflection tables.

That all said, I am not sure it is essential to have the best possible navigation experience for this release. We will likely not come up with something that pleases everyone anyway, so most important is that the user be able to get to all the tables of interest.

**Comment by kirlat** [comment]
I'm thinking that since we have alpheios-project/components#40 implemented now, maybe we can follow the same approach here and show linked tables below the main one in a collapsed state?

I think this will save our users from some confusion of having multiple differently-related tables in several drop-downs (I'm thinking I drop-downs can be overly complex in some cases and simplifying them would be a good thing in general). It will also be consistent with the approach we use for alpheios-project/components#40.

What do you think?

**Comment by balmas** [comment]
I am inclined to request user feedback on the functionality as it currently stands before investing more time in this. 

Would like @monzug to provide her thoughts on it.

**Comment by monzug** [comment]
I am inclined to buy @Kirlat's idea of showing the linked tables in a collapsed state. need to see this visually to have the feeling of it.
going back to  amatum, let's go over what we display right now in build 2.0.3-6 in Chrome:
First, a noun table, next an adj table, then a supine table and last the verb participle table. 
two things here:
1) the order of the tables - as Bridget pointed out: "the order of the tables shown in that navigation should match the order of the lexical entries in the popup"
so we should have verb participle, supine, adj and last noun tables
2) both supine and verb participle do not have the link to the regular verb table or show the regular verb table below. 
I have seen this working, so it's either broken in the current build or missing a fix or something else. Please advise.

**Comment by balmas** [comment]
> 2. both supine and verb participle do not have the link to the regular verb table or show the regular verb table below.
>    I have seen this working, so it's either broken in the current build or missing a fix or something else. Please advise.

@monzug if we add the missing regular verb table in the browser would that be sufficient for now or do you think it's really important to have it in the drop-down?

**Comment by monzug** [comment]
Absolutely. The problem is that with all these issues fixed in different
environment, it's hard to understand what's in and what's out

**Comment by balmas** [comment]
agreed :-)

Ok, so @kirlat if you can look at the ordering, point #1 

>the order of the tables - as Bridget pointed out: "the order of the tables shown in that navigation should match the order of the lexical entries in the popup"
so we should have verb participle, supine, adj and last noun tables

I think that would be the minimum to fix for the release if possible.  But please let me know if this is a lot of work or will require any major changes. Thanks!

**Comment by balmas** [comment]
postponing.

**Comment by monzug** [comment]
e.g.  while testing fix #167 , I come up with this example of drop-down:
lookup for scrutato
it's imperative and verb participle. 
the drop-down is verb and verb participle. under verb I have the imperative.
in the web browser the imperative, infinitive, participle and supine are all under the same section.

**Comment by balmas** [comment]
I'm going to close this as we have had no user requests for changes to the ordering of the Latin inflection tables.
